### DISCUSSION TITLE
Include capability to use Dataverse Search API from ServiceClient

### DISCUSSION THREAD

**Discussion by kheiman61** [microsoft/PowerPlatform-DataverseServiceClient#407] [Ideas]
Maybe there's a way to do this already that I missed, but I'd like to be able to use my existing ServiceClient references in my custom code to also access Dataverse Search instead of having to mix and match ServiceClient and Dataverse Search API calls using direct HTTP/WebAPI calls. I can add some helper methods in my plumbing to assist with Dataverse Search, but then I still have to have multiple authentication approaches which I prefer to avoid.

**Comment by JimDaly** [comment]
@kheiman61 

At around noon today (PST) you should be able find these newly published articles documenting the Search 2.0 API. With this API, you can use the SDK or Web API.

- [Search for Dataverse records](https://learn.microsoft.com/power-apps/developer/data-platform/search/overview)
- [Query](https://learn.microsoft.com/power-apps/developer/data-platform/search/query)
- [Suggest](https://learn.microsoft.com/power-apps/developer/data-platform/search/suggest)
- [Autocomplete](https://learn.microsoft.com/power-apps/developer/data-platform/search/autocomplete)
- [Statistics and Status](https://learn.microsoft.com/power-apps/developer/data-platform/search/statistics-status)
- [Legacy search](https://learn.microsoft.com/power-apps/developer/data-platform/search/legacy)

You will find that we also have C# sample code here:
•	[PowerApps-Samples/dataverse/orgsvc/C#-NETCore/Search/README.md at master · microsoft/PowerApps-Samples (github.com)](https://github.com/microsoft/PowerApps-Samples/blob/master/dataverse/orgsvc/C%23-NETCore/Search/README.md)
•	[PowerApps-Samples/dataverse/webapi/C#-NETx/Search at master · microsoft/PowerApps-Samples (github.com)](https://github.com/microsoft/PowerApps-Samples/tree/master/dataverse/webapi/C%23-NETx/Search#web-api-search-operations-sample)

The C# Web API sample is probably not so useful to you since you can use the SDK ServiceClient with the [pac modelbuilder build command](https://learn.microsoft.com/en-us/power-platform/developer/cli/reference/modelbuilder#pac-modelbuilder-build) and get some level of code generation. These APIs use a lot of JSON strings. The samples for either SDK or Web API include helper classes to serialize/deserialize the objects that you then pass as strings to the API.

Please let me know what you think of the documentation and whether this helps you use your ServiceClient references.

**Reply by kheiman61** [reply]
Thanks Jim! Although for simplicity it would be great if it was fully baked into serviceclient without any code gen, but I get it as Dataverse Search is moving to its own separate API, and may change, have multiple versions, etc. I'm currently already using webapi code with the search endpoint and just rolled my own simple struct/class to hold the search properties, and I'm borrowing the current Auth Token from my existing ServiceClient using the conveniently provided property. Currently this is just POC level so I can revisit the code gen and something more robust. FYI I'm using Dataverse Search API as a 'first pass' query in Semantic Kernel methods for interacting with Dataverse data to do a fuzzy search of data, so I can get a limited dataset back that won't blow up the context of an LLM call when I want to provide some data context to the LLM. I'm mildly concerned that the entire org will throttle to 150 Dataverse Search calls/minute, but will cross that bridge if/when we get there.

**Comment by JimDaly** [comment]
I'll let @MattB-msft  speak to whether this should be fully baked into ServiceClient without any code gen. Because these APIs have so many string properties that expect JSON data, they are hard to use without helper classes. It took some time to work these out in the samples we have provided. Sounds like you had to figure that out yourself too. My wish is that Dataverse would enable a capability to define supporting classes for any custom api request parameters or response properties, then the code gen would provide everything you need.
